### PR TITLE
SourceDataLayer Support

### PR DESCRIPTION
**Pull request by johannes-wolf**
This PR introduces support for a new layer, the `TileSourceDataLayer` and updates `mapget`
to be able to handle layer types other than `TileFeatureLayer`.

### CODE REVIEW

File: libs/model/include/mapget/model/stream.h [hunk 1]
```diff
@@ -36,9 +38,9 @@ class TileLayerStream
     static constexpr Version CurrentProtocolVersion{0, 1, 1};
 
     /** Map to keep track of the highest sent field id per datasource node. */
-    using FieldOffsetMap = std::map<std::string, simfil::FieldId>;
+    using FieldOffsetMap = std::map<std::string, simfil::StringId>;
```
Comment (josephbirkner) [inline libs/model/include/mapget/model/stream.h:41]: rename (grep pls)

File: libs/model/src/info.cpp [hunk 2]
```diff
@@ -313,9 +313,13 @@ nlohmann::json Coverage::toJson() const
 std::shared_ptr<LayerInfo> LayerInfo::fromJson(const nlohmann::json& j, std::string const& layerId)
 {
     try {
+        const auto type = j.value("type", LayerType::Features);
+
         std::vector<FeatureTypeInfo> featureTypes;
-        for (auto& item : j.at("featureTypes")) {
-            featureTypes.push_back(FeatureTypeInfo::fromJson(item));
+        if (type == LayerType::Features) {
```
Comment (josephbirkner) [inline libs/model/src/info.cpp:319]: Idea: Save all SourceDataRefs in a cont. buffer. Each referee stores an index and size -> use simfil::ArrayArena. Maybe create a `TinyFixedSizeArrayArena`, where chunks are only `3B index + 1B length`.

File: libs/model/src/simfilutil.h [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    std::vector<simfil::Value> eval(std::string_view query, simfil::ModelNode const& node)
+    {
+        auto evalFun = [&](auto&& expr) {
+            return simfil::eval(*env_, expr, node);
+        };
+
+        return eval(query, evalFun);
+    }
+
+    std::vector<simfil::Value> eval(std::string_view query, simfil::ModelPool const& pool, size_t rootIndex = 0)
```
Comment (josephbirkner) [inline libs/model/src/simfilutil.h:69]: Please remove this (from simfil as well).

File: libs/model/src/sourcedata.cpp [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+uint32_t SourceDataCompoundNode::size() const
+{
+    if (auto m = object())
+        return m->size();
+    return 0;
+}
+
+ModelNode::Ptr SourceDataCompoundNode::get(const simfil::StringId& field) const
+{
+    if (auto m = object())
```
Comment (josephbirkner) [inline libs/model/src/sourcedata.cpp:73]: Maybe also expose the compound node's `_address` and `_type` fields.

File: libs/model/src/sourcedatalayer.cpp [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    FieldNameResolveFun const& fieldNameResolveFun
+) :
+    TileLayer(in, layerInfoResolveFun),
+    ModelPool(fieldNameResolveFun(nodeId_)),
+    impl_(std::make_unique<Impl>(fieldNameResolveFun(nodeId_)))
+{
+    bitsery::Deserializer<bitsery::InputStreamAdapter> s(in);
+    impl_->readWrite(s);
+    if (s.adapter().error() != bitsery::ReaderError::NoError) {
+        raise(fmt::format(
```
Comment (josephbirkner) [inline libs/model/src/sourcedatalayer.cpp:74]: Use raiseFmt

File: libs/service/include/mapget/service/service.h [hunk 6]
```diff
@@ -60,22 +61,32 @@ class LayerTilesRequest
     std::vector<TileId> tiles_;
 
     /**
-     * The callback function which is called when a result tile is available.
+     * The callback function which is called when all tiles have been processed.
      */
-    std::function<void(TileFeatureLayer::Ptr)> onResult_;
+    std::function<void(RequestStatus)> onDone_;
```
Comment (josephbirkner) [inline libs/service/include/mapget/service/service.h:66]: ?

File: libs/model/include/mapget/model/sourceinfo.h [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+ * is attached to the blob-tree and can be found by looking for the tree node(s)
+ * that match the target blobs SourceRegion.
+ */
+struct SourceDataReference
+{
+    /**
+     * Create a unique layer-id string form a combination of a
+     * layer target and a data-layer id.
+     */
+    static std::string makeLayerId(std::string_view target, uint64_t dataLayerId)
```
Comment (josephbirkner) [inline libs/model/include/mapget/model/sourceinfo.h:84]: This utility function seems a bit out of place in this library :)

File: libs/model/src/info.cpp (hunk 2 above)
Comment (johannes-wolf) [reply libs/model/src/info.cpp:319]: See PR https://github.com/Klebert-Engineering/simfil/pull/90.

**Comment by josephbirkner** [comment]
Please add an override for `setStrings` in `TileSourceDataLayer` (the source data ref qualifier string IDs must be transcoded there).

**Review by josephbirkner — APPROVED** [review]
This looks very nice :+1: And it seems like the effort to minimize the size of the source data references is paying off :smile_cat: 

Two more remarks:

* Wouldn't it be better to store the source data address format on the `TileSourceDataLayer`? 
* We need to update the architecture diagram in the README (it is on Lucid), and mention the existence of the TileSourceDataLayer.

**Pull request merged**